src/coreclr/jit/codegenloongarch64.cpp

```diff
@@ -0,0 +1,9600 @@
+//        result = (prevDimOffset * dimSize) + effectiveIndex
+//    where dimSize is obtained from the arrObj operand
+//
+// Arguments:
+//    arrOffset - the node for which we're generating code
+//
+// Return Value:
+//    None.
+//
+// Notes:
+//    dimSize and effectiveIndex are always non-negative, the former by design,
+//    and the latter because it has been normalized to be zero-based.
+
+void CodeGen::genCodeForArrOffset(GenTreeArrOffs* arrOffset)
+{
+    GenTree*  offsetNode = arrOffset->gtOffset;
+    GenTree*  indexNode  = arrOffset->gtIndex;
+    regNumber tgtReg     = arrOffset->GetRegNum();
+
+    noway_assert(tgtReg != REG_NA);
+
+    if (!offsetNode->IsIntegralConst(0))
+    {
+        emitter*  emit      = GetEmitter();
+        regNumber offsetReg = genConsumeReg(offsetNode);
+        regNumber indexReg  = genConsumeReg(indexNode);
+        regNumber arrReg    = genConsumeReg(arrOffset->gtArrObj);
+        noway_assert(offsetReg != REG_NA);
+        noway_assert(indexReg != REG_NA);
+        noway_assert(arrReg != REG_NA);
+
+        // regNumber tmpReg = arrOffset->GetSingleTempReg();
+
+        unsigned dim    = arrOffset->gtCurrDim;
+        unsigned rank   = arrOffset->gtArrRank;
+        unsigned offset = compiler->eeGetMDArrayLengthOffset(rank, dim);
+
+        // Load tmpReg with the dimension size and evaluate
+        // tgtReg = offsetReg*tmpReg + indexReg.
+        emit->emitIns_R_R_I(INS_ld_w, EA_4BYTE, REG_R21, arrReg, offset);
+        emit->emitIns_R_R_R(INS_mul_d, EA_PTRSIZE, REG_R21, REG_R21, offsetReg);
+        emit->emitIns_R_R_R(INS_add_d, EA_PTRSIZE, tgtReg, REG_R21, indexReg);
+    }
+    else
+    {
+        regNumber indexReg = genConsumeReg(indexNode);
+        if (indexReg != tgtReg)
+        {
+            GetEmitter()->emitIns_R_R_I(INS_ori, emitActualTypeSize(TYP_INT), tgtReg, indexReg, 0);
+        }
+    }
+    genProduceReg(arrOffset);
+}
+
+//------------------------------------------------------------------------
+// genCodeForShift: Generates the code sequence for a GenTree node that
+// represents a bit shift or rotate operation (<<, >>, >>>, rol, ror).
+//
+// Arguments:
+//    tree - the bit shift node (that specifies the type of bit shift to perform).
+//
+// Assumptions:
+//    a) All GenTrees are register allocated.
+//
+void CodeGen::genCodeForShift(GenTree* tree)
+{
+    // var_types   targetType = tree->TypeGet();
+    // genTreeOps  oper       = tree->OperGet();
+    instruction ins  = genGetInsForOper(tree);
+    emitAttr    size = emitActualTypeSize(tree);
+
+    assert(tree->GetRegNum() != REG_NA);
+
+    genConsumeOperands(tree->AsOp());
+
+    GenTree* operand = tree->gtGetOp1();
+    GenTree* shiftBy = tree->gtGetOp2();
+    if (!shiftBy->IsCnsIntOrI())
+    {
+        GetEmitter()->emitIns_R_R_R(ins, size, tree->GetRegNum(), operand->GetRegNum(), shiftBy->GetRegNum());
+    }
+    else
+    {
+        unsigned shiftByImm = (unsigned)s
```

Comment:
Note that this was renamed to `genCall` on other architectures in #58686, so you may wish to do the same renaming to avoid confusion.

Reply:
OK, Thanks.